theano/tensor/elemwise.py

```diff
@@ -1187,12 +1189,12 @@ def _c_all(self, node, nodename, inames, onames, sub):
             for output, oname, i in izip(node.outputs, onames,
                                          range(len(onames))):
                 idtype = output.type.dtype_specs()[1]
+
                 # We alias the scalar variables
                 defines += """
-                #define %(oname)s_i *(%(idtype)s*) (refVar[i%(oname)s])
-                int i%(oname)s;
-                i%(oname)s = outIndexes[%(i)i];
+                #define %(oname)s_i *(t%(i)i*) (refVar[outIndexes[%(i)i]])
```

Comment:
Basically, we would be able to retrieve the type of the variable where the refVar array points to and typedef it to t%(i) so that the casting is properly done (else it will lead to segfaults). However I was not able to find a solution relying only on C code. I think I'm thinking too much object oriented and not enough in C.

That or find an other implementation than macro to alias the inplace variables.

In order to make the test pass I've committed a temporary hack that will be removed from history once a solution is found.